src/WebJobs.Script/Host/FunctionMetadataProvider.cs

```
@@ -0,0 +1,60 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Workers.Rpc;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    internal class FunctionMetadataProvider : IFunctionMetadataProvider
+    {
+        private readonly IEnvironment _environment;
+        private readonly ILogger<FunctionMetadataProvider> _logger;
+        private IWorkerFunctionMetadataProvider _workerFunctionMetadataProvider;
+        private IHostFunctionMetadataProvider _hostFunctionMetadataProvider;
+
+        public FunctionMetadataProvider(ILogger<FunctionMetadataProvider> logger, IWorkerFunctionMetadataProvider workerFunctionMetadataProvider, IHostFunctionMetadataProvider hostFunctionMetadataProvider)
+        {
+            _logger = logger;
+            _workerFunctionMetadataProvider = workerFunctionMetadataProvider;
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
+            _environment = SystemEnvironment.Instance;
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors { get; private set; }
+
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh = false)
+        {
+            bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, _environment);
+
+            if (!workerIndexing)
+            {
+                return await GetMetadataFromHostProvider(workerConfigs, environment, forceRefresh);
+            }
+
+            FunctionMetadataResult functionMetadataResult = await _workerFunctionMetadataProvider.GetFunctionMetadataAsync(workerConfigs, SystemEnvironment.Instance, forceRefresh);
```

We will need a null check here for `_workerFunctionMetadataProvider`.

---

My main point here is - if _workerFunctionMetadataProvider is null, we should throw an exception. We used to do that previously in `AggregateFunctionMetadataProvider` for null check on dispatcher

---

Sorry i caught this very late. The code flow there was actually very different that warranted a null check. We have a lot of code paths calling the method, some with dispatcher being null others not. In this case though it is populated via DI and we almost never expect null here. I will take a look at it regardless